File: ios/Classes/FlutterAsaAttributionPlugin.m
#import "FlutterAsaAttributionPlugin.h"
#import <iAd/iAd.h>
#import <AdServices/AdServices.h>
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#import "AsaTool.h"

@implementation FlutterAsaAttributionPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel methodChannelWithName:@"lianshumin.github.io/flutter_asa_attribution"
                                                                binaryMessenger:[registrar messenger]];
    FlutterAsaAttributionPlugin *instance = [[FlutterAsaAttributionPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"requestAttributionDetails" isEqualToString:call.method]) {
      [AsaTool requestAttributionWithComplete:^(NSDictionary * _Nonnull data, NSError * _Nonnull error) {
          if (data) {
              result(data);
          } else {
              result([FlutterError errorWithCode:@"FAILED" message:error.localizedFailureReason details:nil]);
          }
      }];
  } else if ([@"attributionToken" isEqualToString:call.method]) {
      result([AsaTool attributionToken]);
  } else {
      result(FlutterMethodNotImplemented);
  }
}

@end

File: ios/flutter_asa_attribution/Sources/flutter_asa_attribution/FlutterAsaAttributionPlugin.swift
import Flutter

public class FlutterAsaAttributionPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "lianshumin.github.io/flutter_asa_attribution",
      binaryMessenger: registrar.messenger()
    )
    let instance = FlutterAsaAttributionPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "requestAttributionDetails":
      AsaTool.requestAttribution { data, error in
        if let data = data {
          result(data)
        } else {
          let nsError = error as NSError?
          result(
            FlutterError(
              code: "FAILED",
              message: nsError?.localizedFailureReason ?? nsError?.localizedDescription,
              details: nil
            )
          )
        }
      }
    case "attributionToken":
      result(AsaTool.attributionToken())
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
